csrc/scheduler/reduction.cpp

```diff
@@ -520,32 +554,47 @@ std::shared_ptr<ReductionParams> outerReductionHeuristic(
   const int64_t device_multiprocessor_count =
       (int64_t)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
 
-  auto const max_unroll = ceilDiv(
+  // when unroll in reduction dim and vect in iteration dim, needs to store
+  // loaded data in registers, set a limit on register count we can use to store
+  // unrolled and vectorized data. With 512 threads per block, each thread can
+  // use 128 registers. Assume 64 registers are used to store loaded data from
+  // iter and redu unroll, max_unroll can be up to 128 for fp16.
+  // TODO: investigate why set to 128 registers won't cause register spill when
+  // vectorized by 8 and unrolled by 32.
+  auto max_buffer_register = 64L;
+  auto max_unroll = ceilDiv(
       // Available unrolling based on size of data type
-      (int64_t)16 / (int64_t)max_input_dtype_size,
+      max_buffer_register * scheduler_utils::bytes_per_register /
+          (int64_t)max_input_dtype_size,
       // Reduce unrolling if we have many inputs, start reduction at 4 inputs
       scheduler_utils::lastPow2(
           std::max((int64_t)n_tensor_inputs >> 2, (int64_t)1)));
+  max_unroll = scheduler_utils::lastPow2(max_unroll);
+
+  // avoid using very larege unroll in reduction dim, e.g. when iter dim is
+  // very small, iter_unroll = 1, don't let all unroll go to reduction dim.
+  auto const max_redu_unroll = 64l;
+
+  // set max thread local serial reduction workload
+  // controls max elements block reduce can handle,
+  // which equals: bdimy * redu_unroll * serial
+  auto const max_serial_reduce_top_unroll = 16L;
 
   const int64_t n_elems = total_reduction_numel * total_iteration_numel;
   const int64_t n_waves = 8;
 
-  // if data fits in l2 and we need more parallelization in the iter dim,
-  // we can use a smaller warp size. While thread local data fits in l1, and
-  // iter dim is really small, we can use <32 threads per warp.
-  // TODO: Could get a much more accurate estimation of it the problem fits in
-  // L2
-  const bool fits_in_l2 = n_elems * max_input_dtype_size * n_tensor_inputs <
-      at::cuda::getCurrentDeviceProperties()->l2CacheSize;
-
-  const int64_t min_warp_size = fits_in_l2 ? 16 : 32;
+  // start from small bimdx, leave more space for bdimy to use block reduce
+  const int64_t min_warp_size = 8;
```

Comment:
I renamed `min_warp_size` to `min_bdimx` and added comments:
```
  // [min_bdimx] is used to set [bdimx], which parallels the iteration dim.
  // In CUDA, 32 threads are gouped as a warp, when we set bdimx = 8, the 32
  // threads within a warp access 4 consecutive rows of data from gmem. Assume
  // each row has Y bytes of data and after vectorization each thread accesses
  // 16 bytes of data. So the gmem access pattern is:
  // threads [ 0 to  7] access 128 bytes from address  0 + 0x.
  // threads [ 8 to 15] access 128 bytes from address  Y + 0x.
  // threads [16 to 23] access 128 bytes from address 2Y + 0x.
  // threads [24 to 31] access 128 bytes from address 3Y + 0x.
  // The data is spread into 4 segments, each with 128 bytes.
  // The gmem access is done with 4 128-bytes transactions.
  // If we further reduce [min_bdimx] to 4, the data is spread into 8 rows
  // and we need 8 64-bytes transactions.
  // If we further increase [min_bdimx] to 16, the data is spread into 2
  // rows, each row has 16 x 16 = 256 bytes of data. Since the largest
  // transaction is 128 btyes, we still needs 2 x 2 = 4 128-bytes transactions.
```